### PR TITLE
[fix][OSS] Mixed types & proper buckets + no grad_fn is fine

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [~x] Did you write any new necessary tests?

## What does this PR do?
Fixes https://fb.workplace.com/groups/pytorchLightning/permalink/826203544631599/. I tried for some time but could not find a case in the unit test which would trigger this (inc. with pytorch 1.9 nightly), I've asked the users for repro examples and will update if something comes

Fixes another bug pre-emptively (no report yet), where the buckets could be of the wrong type if these were mixed and bad luck. Adds a unit test for this one

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

cc @ananthsub

### CODE REVIEW

File: fairscale/optim/oss.py [hunk 1]
```diff
@@ -591,13 +591,14 @@ def _setup_flat_buffers(self) -> None:
 
                     # Merge all the trainable params in a single bucket
                     trainable_params = list(filter(lambda x: x.requires_grad, params))
-                    buffer_size = sum(map(lambda x: x.numel(), trainable_params))
-                    bucket = ParamBucket(size=buffer_size, dtype=params[0].dtype, device=device)
+                    if trainable_params:
+                        buffer_size = sum(map(lambda x: x.numel(), trainable_params))
+                        bucket = ParamBucket(size=buffer_size, dtype=trainable_params[0].dtype, device=device)
```
Comment (blefaudeux) [inline fairscale/optim/oss.py:596]: there was a typo here, which could change everything: pick the first param type vs. first trainable param. The trainable params are the ones which go in a bucket, so you could have a config (see added unit test) which could break that, by adding another param of another type which was not trainable, and have it come first in the list

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 2]
```diff
@@ -494,18 +494,24 @@ def _setup_backward_hooks(self) -> None:
...
-            reduce_function = self._get_reduce_fn(index, param, dst_rank)
 
-            self._grad_hooks.append(grad_acc.register_hook(reduce_function))
-            self._grad_accs.append(grad_acc)  # keep this hook in scope
-            self._manual_reduce.append(reduce_function)
+            # See https://pytorch.org/docs/stable/tensors.html?highlight=grad_fn
+            # We're interested in the tensors which will be tracked by Autograd
+            # Some tensors can have gradients independent of the inputs (ie. pooling layer for instance),
+            # these do not need to be sync'ed
+            if p_tmp.grad_fn is not None:
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:503]: the assert here was too strong, we can just ignore the params whose gradients don't depend on the inputs (see link), by definition they don't need to be synced since all the ranks will have the same. I could not repro an issue here though (even with 1.9 nightly), but several users faced the issue with a recent pytorch, so I'm  guessing that pytorch is getting more aggressive at pruning .grad_fn when not required

**Review by ananthsub — APPROVED** [review]

**Review by anj-s — APPROVED** [review]
